### DISCUSSION TITLE
CSS generation location

### DISCUSSION THREAD

**Discussion by GalacticHypernova** [tailwindlabs/tailwindcss#12929] [Help]
I was just wondering where the CSS classes are actually generated? Where does it get all the actual CSS classes and the actual css properties for each class?

**Comment by AM-NiceDay** [comment]
Hey @GalacticHypernova!

In a Tailwind CSS project, the CSS classes are generated in a compiled CSS file, usually `﻿output.css`, `﻿styles.css`, or another name defined by the developer. This generation process occurs when you build your Tailwind CSS through a PostCSS build process.

When you use a class like ﻿`text-lg` in your HTML, Tailwind's PostCSS plugin processes it during the build. It matches this class with a rule found in the configuration file (﻿`tailwind.config.js`, or the defaults if no custom configuration exists).
For ﻿`text-lg`, Tailwind must have a rule that looks something like:
```js
{
  theme: {
    fontSize: {
      lg: '1.125rem',
      // other sizes...
    }
  }
  // other configurations...
}
```

During the build process, Tailwind translates the ﻿text-lg class into a CSS rule like:
```js
.text-lg {
  font-size: 1.125rem;
}
```

This rule is then included in the final compiled CSS output. As a result, when the ﻿text-lg class is used in HTML, it applies a font size of ﻿`1.125rem` to the element.

Here you can find the [Default Tailwind CSS configuration](https://github.com/tailwindlabs/tailwindcss/blob/master/stubs/config.full.js).

At the very beginning of the generated CSS file you can find the default rules applied to "reset" and "normalize" the default browser rules for development convenience. Here's a [Preflight CSS](https://github.com/tailwindlabs/tailwindcss/blob/49cbb414e9df459870c54252c6e93037958d6ac1/src/css/preflight.css). If you find this configuration redundant, you can turn it off in the configuration using `preflight: false` option.

**Reply by GalacticHypernova** [reply]
Hey @AM-NiceDay , thanks for the response!
May I ask where I could find the rule for core classes like flex/grid/etc? (Classes that aren't typically changed)

**Reply by AM-NiceDay** [reply]
Sure, a more complex stuff you can find in the [Core Plugins](https://github.com/tailwindlabs/tailwindcss/blob/49cbb414e9df459870c54252c6e93037958d6ac1/src/corePlugins.js). Here're a couple of links to exact lines:
- [Flex](https://github.com/tailwindlabs/tailwindcss/blob/49cbb414e9df459870c54252c6e93037958d6ac1/src/corePlugins.js#L803)
- [Grid](https://github.com/tailwindlabs/tailwindcss/blob/49cbb414e9df459870c54252c6e93037958d6ac1/src/corePlugins.js#L816)
- [Flex Direction](https://github.com/tailwindlabs/tailwindcss/blob/49cbb414e9df459870c54252c6e93037958d6ac1/src/corePlugins.js#L1237)

**Reply by GalacticHypernova** [reply]
Thank you very much! I was looking everywhere in stubs because that's where the whole configuration was that I forgot to look in the src directory. Thanks again!
